DRAFT - 21.0.0.12 GA release blog (#1910)
OpenLiberty/blogs: posts/2021-11-26-jakarta-ee-9.1.adoc

seo-description: Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers! Several significant bug fixes are also part of this release.
blog_description: Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers! Several significant bug fixes are also part of this release.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Jakarta EE 9.1 support and configurable response headers in Open Liberty 21.0.0.12
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers! Several significant bug fixes are also part of this release.
 
In link:{url-about}[Open Liberty] 21.0.0.12:

* <<jee_9.1, Jakarta EE 9.1 support>>

> * <<response_headers, Configurable Response Headers>>

Reviewer: ```suggestion
* <<response_headers, Configurable Response Headers>>
* <<bugs, Notable bug fixes>>
```

> View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A210012+label%3A%22release+bug%22[21.0.0.12].

Reviewer: Since we already have this link in the #bugs section, lets instead change this to point to the #guides section below.  Something along the lines of 'Along with the updates to the runtime, we've also made updates to our guides'

Reviewer: ```suggestion
Along with the new features and functions added to the runtime, we've also made <<guides, updates to our guides>>.
```

[#run]
== Run your apps using 21.0.0.12

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.12</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.12,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#jee_9.1]
== Jakarta EE 9.1 support

> 

Reviewer: Add the start of this section, can you add the Jakarta EE logo from https://www.eclipse.org/org/artwork/zip_files/jakartaee_color.zip
![image](https://user-images.githubusercontent.com/3076261/143480361-e69a0c47-3034-48d2-8b39-5ec05f628092.png)

> Jakarta EE 9.1 support is now available in Open Liberty!!  This support allows you to run Jakarta EE 9 applications using Java 8, 11, or 17 with other Open Liberty value-add features that are updated to support Jakarta EE 9.  Many of you have followed our progress for delivering Jakarta EE 9.1 via our beta releases and corresponding blogs, and we'd like to thank those that provided feedback along the way.  

Reviewer: ```suggestion
Jakarta EE 9.1 support is now available in Open Liberty!!  This support allows you to run Jakarta EE 9.1 applications using Java 8, 11, or 17 with other Open Liberty value-add features that are updated to support Jakarta EE 9.1.  Many of you have followed our progress for delivering Jakarta EE 9.1 via our beta releases and corresponding blogs, and we'd like to thank those that provided feedback along the way.  
```

If you're targetting a new application for Jakarta EE 9.1, make sure to use the `jakarta` namespace.  For existing applications that you'd like to move from Java EE (and its `javax` namespace) to Jakarta EE (and its `jakarta` namespace) we recommend trying the Eclipse Transformer, an open source project originally developed by members of the Open Liberty team and then contributed to the Eclipse Foundation.  

> With the RESTful Web Services 3.0 (formerly called JAXRS) support in Open Liberty, there is a significant performance improvement for applications who use the RESTful Web Services function.  This performance improvement was achieved by moving our RESTful Web Services implementation from Apache CXF to using RestEasy.  With this new version, CDI is enabled by default, and JSON Binding is not enabled by default with the feature being specified in your `server.xml`.   

Reviewer: ```suggestion
With the RESTful Web Services 3.0 (formerly called JAXRS) support in Open Liberty, there is a significant performance improvement for applications which use the RESTful Web Services function.  This performance improvement was achieved by moving our RESTful Web Services implementation from Apache CXF to using RestEasy.  With this new version, CDI is enabled by default, and JSON Binding is not enabled by default with the feature being specified in your `server.xml`.   
```

> Any Liberty features with API and/or SPI functions that use Jakarta EE APIs as part of their method signatures have been updated to have their package versions be reset to 10.0 when using those API / SPIs with Jakarta EE 9.  Any bundles used for user features that depend on those packages will need to change the import package version range when updating the user feature to use Jakarta EE 9.

Reviewer: ```suggestion
Any Liberty features with API and/or SPI functions that use Jakarta EE APIs as part of their method signatures have been updated to have their package versions be reset to 10.0 when using those API / SPIs with Jakarta EE 9.1.  Any bundles used for user features that depend on those packages will need to change the import package version range when updating the user feature to use Jakarta EE 9.1.
```

> With the introduction of Jakarta EE 9.1, the Jakarta Enterprise Beans 4.0 specification includes a few minor changes over the prior version of the specification, Enterprise JavaBeans (EJB) 3.2, as follows.

Reviewer: ```suggestion
With the introduction of Jakarta EE 9.1, the Jakarta Enterprise Beans 4.0 specification includes a few minor changes over the prior version of the specification, Enterprise JavaBeans (EJB) 3.2, as follows:
```

  
*  Note the new names of the features; all of the same features exist, but the feature name prefix has changed from `ejb` to `enterpriseBeans`.  For example, `enterpriseBeansLite-4.0` is the new version of `ejbLite-3.2`.
* The API package has changed from `javax.ejb` to `jakarta.ejb`
* The `@Schedule` annotation is now repeatable
* The following API methods have been removed:
** `javax.ejb.EJBContext.getCallerIdentity()` -> use `getCallerPrincipal()`
** `javax.ejb.EJBContext.getEnvironment()` -> use `JNDI` lookup in `java:comp/env`
** `javax.ejb.EJBContext.isCallerInRole(java.security.Identity)` -> use `isCallerInRole(String)`
** `javax.ejb.SessionContext.getMessageContext()` (removed with `JAX-RPC`)
   
All other capabilities of Enterprise Beans remain the same as the prior specification version (3.2).

> Enable the Jakarta EE 9 features in your app’s `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9 convenience feature to enable all of the Jakarta EE 9 features at once:

Reviewer: ```suggestion
Enable the Jakarta EE 9.1 features in your app’s `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9.1 convenience feature to enable all of the Jakarta EE 9.1 features at once:
```

[source,xml]
----
<featureManager>
    <feature>jakartaee-9.1</feature>
</featureManager>
----

> Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9 Web Profile features at once:

Reviewer: ```suggestion
Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9.1 Web Profile features at once:
```

[source,xml]
----
<featureManager>
    <feature>webProfile-9.1</feature>
</featureManager>
----

|=======================

> |Jakarta EE 8 feature name & short name	|Jakarta EE 9 feature name & short name

Reviewer: ```suggestion
|Jakarta EE 8 feature name & short name	|Jakarta EE 9.1 feature name & short name
```

|Java Web Services (jaxws-2.2) |Jakarta XML Web Services 3.0 (xmlWS-3.0)
|Java Connector Architecture (jca-1.7) |Jakarta Connectors 2.0 (connectors-2.0)
|Java Connector Architecture Security Inflow (jcaInboundSecurity-1.0) |Jakarta Connectors Inbound Security 2.0 (connectorsInboundSecurity-2.0)
|Java Message Service (jms-2.0) |Jakarta Messaging 3.0 (messaging-3.0)
|Java Persistence API (jpa-2.2) |Jakarta Persistence 3.0 (persistence-3.0)
|Java Persistence API Container (jpaContainer-2.2) |Jakarta Persistence Container 3.0 (persistenceContainer-3.0)
|JavaServer Faces (jsf-2.3) |Jakarta Server Faces 3.0 (faces-3.0)
|JavaServer Faces Container (jsfContainer-2.3) |Jakarta Server Faces Container 3.0 (facesContainer-3.0)
|JavaServer Pages (jsp-2.3) |Jakarta Server Pages 3.0 (pages-3.0)
|JMS Client for Message Server (wasJmsClient-2.0) |Messaging Server Client 3.0 (messagingClient-3.0)
|Message Server Security (wasJmsSecurity-1.0) |Messaging Server Security 3.0 (messagingSecurity-3.0)
|Message Server (wasJmsServer-1.0) |Messaging Server 3.0 (messagingServer-3.0)
|=======================

> For more information, see the link:https://staging-openlibertyio.mybluemix.net/docs/21.0.0.12/reference/jakarta-ee9-feature-updates.html[Jakarta EE 9 feature updates] documentation.

Reviewer: ```suggestion
For more information, see the link:{url-prefix}/docs/latest/reference/jakarta-ee9-feature-updates.html[Jakarta EE 9.1 feature updates] documentation.
```

> For details regarding the APIs and SPIs, check out the link:https://staging-openlibertyio.mybluemix.net/docs/21.0.0.12/reference/javadoc/liberty-jakartaee9.1-javadoc.html[Jakarta EE 9.1 javadoc].

Reviewer: ```suggestion
For details regarding the APIs and SPIs, check out the link:{url-prefix}/docs/latest/reference/javadoc/liberty-jakartaee9.1-javadoc.html[Jakarta EE 9.1 javadoc].
```

@Schema
public class Example {
    private String field1;
    public String field2;
}
----
+
results in
+
[source, yaml]
----
  schemas:
    Example:
      type: object
      properties:
        field1:
          type: string
        field2:
          type: string
----
+
The field `field1` should not have appeared in the generated OpenAPI document as it is private. This issue has been fixed by setting the `mp.openapi.extensions.smallrye.private-properties.enable` property to disable non-public properties by default.

* link:https://github.com/OpenLiberty/open-liberty/issues/19057[Port bind skipped at server startup]
+
Previously, in an extremely rare scenario, configured ports could silently fail to bind - preventing Liberty from using them. This issue was caused by a subtle race condition in the code responsible for delaying the port bind until the server is ready to handle traffic.
+
In the failing scenario, the port started message would not be emitted - for example the following message would be missing:
+
`CWWKO0219I: TCP Channel defaultHttpEndpoint has been started and is now listening for requests on host *  (IPv4) port 9080.`
+
and the following FFDC will be seen:
+   

> [source]

Reviewer: ```suggestion
[source, no_copy]
```

* link:https://github.com/OpenLiberty/open-liberty/issues/19197[ClassCastException in JSP relating to JDT internal classes]
+
Open Liberty 21.0.0.9 introduced a bug where the following error occured for certain class lookups in JSP:
+
[source]
----
Error 500: java.lang.ClassCastException: class org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding cannot be cast to class org.eclipse.jdt.internal.compiler.lookup.TypeBinding (org.eclipse.jdt.internal.compiler.lookup.PlainPackageBinding and org.eclipse.jdt.internal.compiler.lookup.TypeBinding are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @3522bc53)
----
+
This issue has now been fixed, as the private `isPackage(String)` method needed to be updated to correctly identify that `org.apache.jsp.test_jsp$1InnerClass` is a class and not a package.

> *New and updated guides since the previous release*

Reviewer: ```suggestion
[#guides]
== New and updated guides since the previous release
```

> As Open Liberty features and functionality continue to grow, we continue to add link:https://openliberty.io/guides/?search=new&key=tag[new guides] to link:http://openliberty.io/[openliberty.io] on those topics to make their adoption as easy as possible.  Existing guides also receive updates in order to address any reported bugs/issues, keep the content current to ensure it doesn't grow stale, and sometimes to expand what's covered under the topic.

Reviewer: ```suggestion
As Open Liberty features and functionality continue to grow, we continue to add link:https://openliberty.io/guides/?search=new&key=tag[new guides] to link:http://openliberty.io/[openliberty.io] on those topics to make their adoption as easy as possible.  Existing guides also receive updates in order to address any reported bugs/issues, keep their content current, and expand what their topic covers.
```

> * link:https://openliberty.io/guides/maven-multimodules.html[Creating a multi-module application]

Reviewer: ```suggestion
* link:{url-prefix}/guides/maven-multimodules.html[Creating a multi-module application]
```

= Jakarta EE 9.1 support and configurable response headers in Open Liberty 21.0.0.12
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 21.0.0.12 offers exciting new enhancements such as Jakarta EE 9.1 support, and configurable response headers, which offer more granular control over response headers. This release also provides several significant bug fixes.

Reviewer: Jakarta EE 9.1 support is now available as part of Open Liberty 21.0.0.12, alongside configurable response headers, which offer more granular control over response headers!  Several significant bug fixes are also part of this release.

 
In link:{url-about}[Open Liberty] 21.0.0.12:

* <<jee_9.1, Jakarta EE 9.1 support>>
* <<response_headers, Configurable Response Headers>>

> View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21.0.0.12+label%3A%22release+bug%22[21.0.0.12].

Reviewer: link is wrong. this seems to happen frequently for the release blog posts.... can you update the template so that the link doesn't use RELEASE_VERSION and makes it clear that the version is of the X00X format (no periods)

> We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21.0.0.12+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.12].

Reviewer: link is wrong.  this seems to happen frequently for the release blog posts.... can you update the template so that the link doesn't use `RELEASE_VERSION` and makes it clear that the version is of the X00X format (no periods)
